docs/debug-port.md

```diff
@@ -0,0 +1,108 @@
+# `cloud-hypervisor` debug IO port
+
+`cloud-hypervisor` uses the [`0x80`](https://www.intel.com/content/www/us/en/support/articles/000005500/boards-and-kits.html)
+I/O port to trace user defined guest events.
+
+Whenever the guest write one byte between `0x0` and `0xF` on this particular I/O port,
+`cloud-hypervisor` will log and timestamp that event at the `debug` log level.
+
+It is up to the guest stack to decide when and what to write to the 0x80 port
+in order to signal the host about specific events and have `cloud-hypervisor`
+log it.
+
+There are 3 pre-defined debug port codes. When logging a write of one of those
+codes to the debug port, `cloud-hypervisor` will add a pre-defined string to the
+logs.
+
+| Code  | Event                    | Log string         |
+| ----- | ------------------------ | ------------------ |
+| `0x0` | Firmware starting point  | `Firmware START`   |
+| `0x1` | Kernel starting point    | `Kernel START`     |
+| `0x2` | Userspace starting point | `Userspace START`  |
+
+One typical use case is guest boot time measurement. By writing different values
+to the debug I/O port at different steps during the boot process: when the
+firmware starts, when the kernel is loaded, when userspace starts, etc. By using
+different values to write on the debug port, `cloud-hypervisor` will log and
+timestamp all those stepss. That provides a basic but convenient way of measuring
+not only the overall guest boot time but all intermediate steps as well.
+
+Below is an example of how to instrument the hypervisor firwmare in order to use
+the `cloud-hypervisor` debug port to time and trace the firmware boot process.
+
+## Firmware boot process tracing
+
+### Rust hypervisor firmware I/O port writes
+
+First of all we need to patch the
+[Rust hypervisor firmware](https://github.com/intel/rust-hypervisor-firmware)
+code in order to place our trace points appropriately:
+
+```
+diff --git a/src/loader.rs b/src/loader.rs
```

Comment:
Rather than include a diff in here I can just update the firmware and spin out a new binary.